ts/index.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * Inject raw response, so we can know if middleware has responsed.
+ */
+function makeInjectedResponse(koaCtx, /*markHandled,*/ whenEnded) {
+  let res = koaCtx.res;
+
+  res.on('close', whenEnded).on('finish', whenEnded);
+
+  let dummyRes = Object.create(res);
+  [
+    'setHeader',
+    'writeHead',
+    'write',
+    'end'
+  ].forEach(name => {
+    dummyRes[name] = function (...args) {
+      res[name](...args);
+      // koa2.0 initial assign statusCode to 404, reset to 200
+      if (res.statusCode === 404) {
+        res.statusCode = 200;
```

Comment:
What if `writeHead` set the status code to `404` intentionally? Is this a valid assumption? Does it make a difference if it comes before calling the corresponding method on `res`?

Maybe we replace `404` with our own number or flag, check if it's equal to that, set it accordingly?